File: lintcode/trapping_rain_water.py
class Solution:
    # @param heights: a list of integers
    # @return: a integer
    def trapRainWater(self, heights):
        h, left, right, result = 0, 0, len(heights) - 1, 0
        while left <= right:
            h = max(h, min(heights[left], heights[right]))
            if heights[left] < heights[right]:
                if heights[left] < h: result += h - heights[left]
                left += 1
            else:
                if heights[right] < h: result += h - heights[right]
                right -= 1
        return result

File: leetcode/trapping_rain_water.js
/* Given n non-negative integers representing an elevation map where the width 
 * of each bar is 1, compute how much water it is able to trap after raining.
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    var result = 0
    for (var h = 0, left = 0, right = height.length - 1; left <= right; ) {
        h = Math.max(h, Math.min(height[left], height[right]))
        if (height[left] < height[right]) {
            if (height[left] < h)
                result += h - height[left]
            ++left
        } else {
            if (height[right] < h)
                result += h - height[right]
            --right
        }
    }
    return result
};
